File: problems/1504.py
class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        n = len(mat)
        m = len(mat[0])
        prs_c = [[ 0 for _ in range(m+1)] for _ in range(n+1)]

        ans = 0
        for i in range(1,n+1):
            for j in range(1,m+1):
                if mat[i-1][j-1] == 0:
                    prs_c[i][j] = 0
                    continue      
                prs_c[i][j] = 1 + prs_c[i-1][j]
        
        for i in range(1,n+1):
            for j in range(1,m+1):
                min_n = 1000
                for k in range(j,m+1):
                    min_n = min(min_n,prs_c[i][k])
                    ans += min_n

        return ans

File: problems/1504.cpp
class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        int m = mat[0].size();
        vector<int> heights(m,0);
        int ans = 0;
        for(const auto& row : mat){
            for(int i=0;i<m;i++)
                heights[i] = (row[i]==0)? 0: heights[i]+1;

            stack<vector<int>> st;
            st.push({-1,0,-1}); // pos cur h
            for(int i=0;i<m;i++){
                int h = heights[i];
                while( st.top()[2] >= h )
                    st.pop();
                
                auto& top = st.top();
                int p = top[0];
                int prev = top[1];
                int cur = prev + (i-p)*h;
                ans += cur;
                st.push({i,cur,h});
            }
        }
        return ans;
    }
};
